File: ios/Sources/ACRendering/Views/AdaptiveCardView.swift
// Copyright (c) Microsoft Corporation. All rights reserved.
// Author: Vikrant Singh (github.com/VikrantSingh01)
// Licensed under the MIT License.

import SwiftUI
import ACCore
import ACActions
import ACAccessibility
import ACInputs

/// Root SwiftUI view for rendering an Adaptive Card
public struct AdaptiveCardView: View {
    let cardJson: String?
    let preParsedCard: AdaptiveCard?
    let templateData: [String: Any]?
    let hostConfig: HostConfig
    let configuration: CardConfiguration?
    let actionDelegate: ActionDelegate?
    let onCardParsed: ((AdaptiveCard) -> Void)?
    let onCardParseError: ((Error) -> Void)?
    var pendingActionTitle: Binding<String?>?
    let onRefreshNeeded: ((CardAction) -> Void)?

    @StateObject private var viewModel = CardViewModel()
    @StateObject private var validationState = ValidationState()
    @State private var cardWidth: CGFloat = 0
    private let actionHandler: ActionHandler

    // MARK: - New API (Phase 2)

    /// Creates an Adaptive Card view from a pre-parsed card and configuration.
    ///
    /// ```swift
    /// let result = AdaptiveCards.parse(jsonString)
    /// if let card = result.card {
    ///     AdaptiveCardView(card: card, configuration: .teams(theme: .dark))
    /// }
    /// ```
    public init(
        card: AdaptiveCard,
        configuration: CardConfiguration = .default,
        onRefreshNeeded: ((CardAction) -> Void)? = nil
    ) {
        self.preParsedCard = card
        self.cardJson = nil
        self.templateData = nil
        self.hostConfig = configuration.hostConfig
        self.configuration = configuration
        self.actionDelegate = nil
        self.actionHandler = DefaultActionHandler()
        self.pendingActionTitle = nil
        self.onCardParsed = nil
        self.onCardParseError = nil
        self.onRefreshNeeded = onRefreshNeeded
    }

    /// Creates an Adaptive Card view from a JSON string and configuration.
    ///
    /// ```swift
    /// AdaptiveCardView(json: jsonString, configuration: .teams(theme: .dark))
    /// ```
    public init(
        json: String,
        data: [String: Any]? = nil,
        configuration: CardConfiguration = .default,
        onRefreshNeeded: ((CardAction) -> Void)? = nil
    ) {
        self.cardJson = json
        self.preParsedCard = nil
        self.templateData = data
        self.hostConfig = configuration.hostConfig
        self.configuration = configuration
        self.actionDelegate = nil
        self.actionHandler = DefaultActionHandler()
        self.pendingActionTitle = nil
        self.onCardParsed = nil
        self.onCardParseError = nil
        self.onRefreshNeeded = onRefreshNeeded
    }

    // MARK: - Legacy API (kept for backward compatibility)

    /// Creates an Adaptive Card view (legacy API).
    /// - Parameters:
    ///   - cardJson: The card JSON string (may contain `${expression}` template syntax)
    ///   - templateData: Optional data context for template expansion
    ///   - hostConfig: Host configuration for theming and layout
    ///   - actionDelegate: Delegate for handling card actions
    ///   - actionHandler: Internal action dispatcher
    ///   - pendingActionTitle: Binding to trigger an action by title (for test automation)
    ///   - onCardParsed: Called when the card is successfully parsed
    ///   - onCardParseError: Called when card parsing fails
    public init(
        cardJson: String,
        templateData: [String: Any]? = nil,
        hostConfig: HostConfig = TeamsHostConfig.create(),
        actionDelegate: ActionDelegate? = nil,
        actionHandler: ActionHandler = DefaultActionHandler(),
        pendingActionTitle: Binding<String?>? = nil,
        onCardParsed: ((AdaptiveCard) -> Void)? = nil,
        onCardParseError: ((Error) -> Void)? = nil
    ) {
        self.cardJson = cardJson
        self.preParsedCard = nil
        self.templateData = templateData
        self.hostConfig = hostConfig
        self.configuration = nil
        self.actionDelegate = actionDelegate
        self.actionHandler = actionHandler
        self.pendingActionTitle = pendingActionTitle
        self.onCardParsed = onCardParsed
        self.onCardParseError = onCardParseError
        self.onRefreshNeeded = nil
    }

    public var body: some View {
        content
            .environmentObject(viewModel)
            .environment(\.hostConfig, hostConfig)
            .environment(\.actionDelegate, actionDelegate)
            .environment(\.actionHandler, actionHandler)
            .environment(\.validationState, validationState)
            .environment(\.featureFlags, configuration?.featureFlags ?? FeatureFlags())
            .onAppear {
                if let preParsedCard = preParsedCard {
                    // Pre-parsed card — set directly without parsing
                    viewModel.card = preParsedCard
                } else if let cardJson = cardJson {
                    viewModel.parseCard(json: cardJson, templateData: templateData)
                }
            }
            .onChange(of: cardJson) { newJson in
                if let newJson = newJson {
                    viewModel.parseCard(json: newJson, templateData: templateData)
                }
            }
            .onChange(of: viewModel.card) { card in
                if let card = card {
                    onCardParsed?(card)
                    // Handle pending action that arrived before the card was parsed
                    if let title = pendingActionTitle?.wrappedValue {
                        DispatchQueue.main.asyncAfter(deadline: .now() + 0.3) {
                            self.triggerAction(byTitle: title, in: card)
                            self.pendingActionTitle?.wrappedValue = nil
                        }
                    }
                }
            }
            .onChange(of: viewModel.parsingErrorId) { _ in
                if let error = viewModel.parsingError {
                    onCardParseError?(error)
                }
            }
            .onChange(of: pendingActionTitle?.wrappedValue) { title in
                guard let title = title, let card = viewModel.card else { return }
                triggerAction(byTitle: title, in: card)
                pendingActionTitle?.wrappedValue = nil
            }
    }

    /// Finds an action by title in the card and triggers it via the action handler.
    @MainActor
    private func triggerAction(byTitle title: String, in card: AdaptiveCard) {
        let allActions = Self.collectAllActions(from: card)
        if let action = allActions.first(where: { $0.title == title }) {
            actionHandler.handle(action, delegate: actionDelegate, viewModel: viewModel)
        }
    }

    /// Recursively collects all actions from the card (card-level + body ActionSets).
    private static func collectAllActions(from card: AdaptiveCard) -> [CardAction] {
        var actions: [CardAction] = card.actions ?? []
        if let body = card.body {
            actions.append(contentsOf: collectActionsFromElements(body))
        }
        return actions
    }

    private static func collectActionsFromElements(_ elements: [CardElement]) -> [CardAction] {
        var actions: [CardAction] = []
        for element in elements {
            if case .actionSet(let actionSet) = element {
                actions.append(contentsOf: actionSet.actions)
            }
            // Recurse into containers
            if case .container(let container) = element, let items = container.items {
                actions.append(contentsOf: collectActionsFromElements(items))
            }
            if case .columnSet(let columnSet) = element {
                for column in columnSet.columns {
                    if let items = column.items {
                        actions.append(contentsOf: collectActionsFromElements(items))
                    }
                }
            }
        }
        return actions
    }

    @ViewBuilder
    private var content: some View {
        if let error = viewModel.parsingError {
            errorView(error: error)
        } else if let card = viewModel.card {
            cardContent(card: card)
        } else {
            ProgressView()
        }
    }

    private func cardContent(card: AdaptiveCard) -> some View {
        let showDiagnostics = configuration?.diagnosticsEnabled == true

        return VStack(spacing: 0) {
            if let body = card.body, !body.isEmpty {
                ForEach(Array(body.enumerated()), id: \.element.id) { index, element in
                    if viewModel.isElementVisible(elementId: element.elementId) {
                        ElementView(element: element, hostConfig: hostConfig)
                            .padding(.top, index > 0 ? Self.spacingValue(for: element.spacing, hostConfig: hostConfig) : 0)
                    }
                }
            }

            if let actions = card.actions, !actions.isEmpty {
                ActionSetView(actions: actions, hostConfig: hostConfig)
                    .padding(.top, CGFloat(hostConfig.spacing.default))
            }
        }
        .frame(minWidth: 0, maxWidth: .infinity, alignment: .leading)
        .clipped()
        .padding(CGFloat(hostConfig.spacing.padding))
        .fixedSize(horizontal: false, vertical: true)
        .background(Color(hex: hostConfig.containerStyles.default.backgroundColor))
        .background(
            GeometryReader { geometry in
                Color.clear.preference(key: CardWidthPreferenceKey.self, value: geometry.size.width)
            }
        )
        .onPreferenceChange(CardWidthPreferenceKey.self) { width in
            if width > 0 { cardWidth = width }
        }
        .environment(\.widthCategory, WidthCategory.from(width: cardWidth, hostConfig: hostConfig))
        .environment(\.layoutDirection, card.rtl == true ? .rightToLeft : .leftToRight)
        .overlay(alignment: .topTrailing) {
            if showDiagnostics {
                DiagnosticsOverlayView(
                    card: card,
                    parseTimeMs: viewModel.lastParseTimeMs
                )
                .padding(4)
            }
        }
        .task(id: card.refresh?.expires) {
            // Auto-refresh: schedule callback when card expires
            guard let onRefreshNeeded = onRefreshNeeded,
                  let refresh = card.refresh,
                  let expiresString = refresh.expires else { return }
            let formatter = ISO8601DateFormatter()
            formatter.formatOptions = [.withInternetDateTime, .withFractionalSeconds]
            guard let expiresDate = formatter.date(from: expiresString) ?? ISO8601DateFormatter().date(from: expiresString) else { return }
            let delay = expiresDate.timeIntervalSinceNow
            if delay > 0 {
                try? await Task.sleep(nanoseconds: UInt64(delay * 1_000_000_000))
                onRefreshNeeded(refresh.action)
            } else {
                // Already expired — notify immediately
                onRefreshNeeded(refresh.action)
            }
        }
    }

    private static func spacingValue(for spacing: Spacing?, hostConfig: HostConfig) -> CGFloat {
        guard let spacing = spacing else {
            return CGFloat(hostConfig.spacing.default)
        }
        switch spacing {
        case .none: return 0
        case .extraSmall: return 4
        case .small: return CGFloat(hostConfig.spacing.small)
        case .default: return CGFloat(hostConfig.spacing.default)
        case .medium: return CGFloat(hostConfig.spacing.medium)
        case .large: return CGFloat(hostConfig.spacing.large)
        case .extraLarge: return CGFloat(hostConfig.spacing.extraLarge)
        case .padding: return CGFloat(hostConfig.spacing.padding)
        }
    }

    private func errorView(error: Error) -> some View {
        VStack(spacing: 8) {
            Image(systemName: "exclamationmark.triangle")
                .font(.largeTitle)
                .foregroundColor(.red)
            Text("Failed to parse card")
                .font(.headline)
            Text(error.localizedDescription)
                .font(.caption)
                .foregroundColor(.secondary)
        }
        .padding()
    }
}

// MARK: - Preference Keys

private struct CardWidthPreferenceKey: PreferenceKey {
    static var defaultValue: CGFloat = 0
    static func reduce(value: inout CGFloat, nextValue: () -> CGFloat) {
        value = max(value, nextValue())
    }
}

// MARK: - Environment Keys

private struct HostConfigKey: EnvironmentKey {
    static let defaultValue: HostConfig = TeamsHostConfig.create()
}

private struct ActionDelegateKey: EnvironmentKey {
    static let defaultValue: ActionDelegate? = nil
}

private struct ActionHandlerKey: EnvironmentKey {
    static let defaultValue: ActionHandler = DefaultActionHandler()
}

private struct ValidationStateKey: EnvironmentKey {
    static let defaultValue: ValidationState = ValidationState()
}

private struct WidthCategoryKey: EnvironmentKey {
    static let defaultValue: WidthCategory = .narrow
}

private struct FeatureFlagsKey: EnvironmentKey {
    static let defaultValue: FeatureFlags = FeatureFlags()
}

extension EnvironmentValues {
    /// Current card width category for targetWidth responsive filtering.
    public var widthCategory: WidthCategory {
        get { self[WidthCategoryKey.self] }
        set { self[WidthCategoryKey.self] = newValue }
    }

    var hostConfig: HostConfig {
        get { self[HostConfigKey.self] }
        set { self[HostConfigKey.self] = newValue }
    }

    var actionDelegate: ActionDelegate? {
        get { self[ActionDelegateKey.self] }
        set { self[ActionDelegateKey.self] = newValue }
    }

    var actionHandler: ActionHandler {
        get { self[ActionHandlerKey.self] }
        set { self[ActionHandlerKey.self] = newValue }
    }

    var validationState: ValidationState {
        get { self[ValidationStateKey.self] }
        set { self[ValidationStateKey.self] = newValue }
    }

    /// Feature flags for fallback/requires evaluation
    public var featureFlags: FeatureFlags {
        get { self[FeatureFlagsKey.self] }
        set { self[FeatureFlagsKey.self] = newValue }
    }
}

File: android/ac-rendering/src/main/kotlin/com/microsoft/adaptivecards/rendering/composables/AdaptiveCardView.kt
// Copyright (c) Microsoft Corporation. All rights reserved.
// Author: Vikrant Singh (github.com/VikrantSingh01)
// Licensed under the MIT License.

package com.microsoft.adaptivecards.rendering.composables

import androidx.compose.foundation.background
import androidx.compose.foundation.layout.BoxWithConstraints
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.rememberScrollState
import androidx.compose.foundation.shape.RoundedCornerShape
import androidx.compose.foundation.verticalScroll
import androidx.compose.material3.MaterialTheme
import androidx.compose.material3.Text
import androidx.compose.runtime.Composable
import androidx.compose.runtime.CompositionLocalProvider
import androidx.compose.runtime.LaunchedEffect
import androidx.compose.runtime.MutableState
import androidx.compose.runtime.collectAsState
import androidx.compose.runtime.compositionLocalOf
import androidx.compose.runtime.getValue
import androidx.compose.runtime.remember
import androidx.compose.ui.Modifier
import androidx.compose.ui.platform.LocalDensity
import androidx.compose.ui.unit.dp
import androidx.lifecycle.viewmodel.compose.viewModel
import com.microsoft.adaptivecards.core.hostconfig.HostConfig
import com.microsoft.adaptivecards.core.models.*
import com.microsoft.adaptivecards.rendering.theme.HostConfigProvider
import com.microsoft.adaptivecards.rendering.modifiers.adaptiveSeparator
import com.microsoft.adaptivecards.rendering.modifiers.adaptiveSpacing
import com.microsoft.adaptivecards.rendering.modifiers.SeparatorLine
import com.microsoft.adaptivecards.charts.BarChartView
import com.microsoft.adaptivecards.charts.DonutChartView
import com.microsoft.adaptivecards.charts.LineChartView
import com.microsoft.adaptivecards.charts.PieChartView
import com.microsoft.adaptivecards.rendering.registry.GlobalElementRendererRegistry
import com.microsoft.adaptivecards.core.CardConfiguration
import com.microsoft.adaptivecards.rendering.viewmodel.ActionHandler
import com.microsoft.adaptivecards.rendering.viewmodel.CardViewModel
import com.microsoft.adaptivecards.rendering.viewmodel.DefaultActionHandler
import com.microsoft.adaptivecards.accessibility.RTLSupport
import com.microsoft.adaptivecards.core.models.WidthCategory
import com.microsoft.adaptivecards.core.models.shouldShowForTargetWidth
import com.microsoft.adaptivecards.core.models.targetWidth

/**
 * CompositionLocal providing the current card width category for targetWidth filtering.
 * Defaults to Narrow (typical phone width).
 */
val LocalWidthCategory = compositionLocalOf { WidthCategory.Narrow }

/**
 * CompositionLocal providing the feature flags for fallback/requires evaluation.
 */
val LocalFeatureFlags = compositionLocalOf { com.microsoft.adaptivecards.core.FeatureFlags() }

/**
 * CompositionLocal providing the CardViewModel to input renderers registered via the
 * GlobalElementRendererRegistry. This avoids a circular dependency between ac-rendering
 * and ac-inputs while allowing host apps to wire up actual input composables.
 */
val LocalCardViewModel = compositionLocalOf<CardViewModel?> { null }

/**
 * Main entry point for rendering an Adaptive Card
 *
 * @param cardJson The JSON string of the adaptive card (may contain `${expression}` template syntax)
 * @param templateData Optional data context for template expansion
 * @param hostConfig Optional host configuration
 * @param actionHandler Handler for card actions
 * @param modifier Modifier for the card container
 * @param viewModel Optional ViewModel for state management
 * @param pendingActionTitle Mutable state to trigger an action by title (for test automation)
 * @param onCardParsed Called when the card is successfully parsed
 * @param onCardParseError Called when card parsing fails
 */
@Composable
fun AdaptiveCardView(
    cardJson: String,
    templateData: Map<String, Any?>? = null,
    hostConfig: HostConfig? = null,
    actionHandler: ActionHandler = DefaultActionHandler(),
    modifier: Modifier = Modifier,
    viewModel: CardViewModel = viewModel(),
    pendingActionTitle: MutableState<String?>? = null,
    onCardParsed: ((AdaptiveCard) -> Unit)? = null,
    onCardParseError: ((String) -> Unit)? = null
) {
    val card by viewModel.card.collectAsState()
    val parseError by viewModel.parseError.collectAsState()

    LaunchedEffect(cardJson, templateData) {
        viewModel.parseCard(cardJson, templateData)
    }

    // Lifecycle callbacks
    LaunchedEffect(card) {
        card?.let { onCardParsed?.invoke(it) }
    }
    LaunchedEffect(parseError) {
        parseError?.let { onCardParseError?.invoke(it) }
    }

    // Handle pending action trigger (for test automation deep links)
    val pendingTitle = pendingActionTitle?.value
    LaunchedEffect(pendingTitle, card) {
        if (pendingTitle != null && card != null) {
            val allActions = collectAllActions(card)
            val matchingAction = allActions.firstOrNull { it.title == pendingTitle }
            if (matchingAction != null) {
                handleAction(matchingAction, actionHandler, viewModel)
            }
            pendingActionTitle?.value = null
        }
    }

    // Display error if parsing failed
    parseError?.let { errorMessage ->
        androidx.compose.material3.Text(
            text = "Failed to render card: $errorMessage",
            color = androidx.compose.material3.MaterialTheme.colorScheme.error,
            modifier = modifier
        )
        return
    }

    card?.let { adaptiveCard ->
        HostConfigProvider(hostConfig = hostConfig ?: com.microsoft.adaptivecards.core.hostconfig.HostConfigParser.default()) {
            RTLSupport(isRTL = adaptiveCard.rtl == true) {
                BoxWithConstraints(modifier = modifier.fillMaxWidth()) {
                    val density = LocalDensity.current
                    val hc = com.microsoft.adaptivecards.rendering.theme.LocalHostConfig.current
                    val widthDp = with(density) { constraints.maxWidth.toDp().value }
                    val widthCategory = WidthCategory.fromDp(
                        widthDp,
                        veryNarrowBreakpoint = hc.hostWidth.veryNarrow,
                        narrowBreakpoint = hc.hostWidth.narrow,
                        standardBreakpoint = hc.hostWidth.standard
                    )

                    CompositionLocalProvider(
                        LocalWidthCategory provides widthCategory,
                        LocalCardViewModel provides viewModel
                    ) {
                        Column(
                            modifier = Modifier
                                .fillMaxWidth()
                                .verticalScroll(rememberScrollState())
                        ) {
                            // Render body elements
                            adaptiveCard.body?.forEachIndexed { index, element ->
                                RenderElement(
                                    element = element,
                                    isFirst = index == 0,
                                    viewModel = viewModel,
                                    actionHandler = actionHandler
                                )
                            }

                            // Render actions
                            adaptiveCard.actions?.let { actions ->
                                if (actions.isNotEmpty()) {
                                    ActionSetView(
                                        actions = actions,
                                        actionHandler = actionHandler,
                                        viewModel = viewModel,
                                        modifier = Modifier.fillMaxWidth()
                                    )
                                }
                            }
                        }
                    }
                }
            }
        }
    }
}

/**
 * New API (Phase 2): Render a pre-parsed Adaptive Card with a CardConfiguration.
 *
 * ```kotlin
 * val result = AdaptiveCards.parse(jsonString)
 * result.card?.let { card ->
 *     AdaptiveCardView(
 *         card = card,
 *         configuration = CardConfiguration.teams(TeamsTheme.Dark),
 *         onAction = { event -> handleAction(event) }
 *     )
 * }
 * ```
 */
@Composable
fun AdaptiveCardView(
    card: AdaptiveCard,
    configuration: CardConfiguration = CardConfiguration.Default,
    actionHandler: ActionHandler = DefaultActionHandler(),
    modifier: Modifier = Modifier,
    viewModel: CardViewModel = viewModel(),
    onCardParsed: ((AdaptiveCard) -> Unit)? = null,
    onRefreshNeeded: ((CardAction) -> Unit)? = null
) {
    // Set the pre-parsed card directly
    LaunchedEffect(card) {
        viewModel.setCard(card)
        onCardParsed?.invoke(card)
    }

    // Auto-refresh: schedule callback when card expires
    if (onRefreshNeeded != null && card.refresh?.expires != null) {
        LaunchedEffect(card.refresh?.expires) {
            val expiresStr = card.refresh?.expires ?: return@LaunchedEffect
            val refreshAction = card.refresh?.action ?: return@LaunchedEffect
            try {
                val expiresInstant = java.time.Instant.parse(expiresStr)
                val delayMs = java.time.Duration.between(java.time.Instant.now(), expiresInstant).toMillis()
                if (delayMs > 0) {
                    kotlinx.coroutines.delay(delayMs)
                }
                onRefreshNeeded(refreshAction)
            } catch (_: Exception) {
                // Invalid date format — ignore
            }
        }
    }

    val currentCard by viewModel.card.collectAsState()

    currentCard?.let { adaptiveCard ->
        HostConfigProvider(hostConfig = configuration.hostConfig) {
            RTLSupport(isRTL = adaptiveCard.rtl == true) {
                BoxWithConstraints(modifier = modifier.fillMaxWidth()) {
                    val density = LocalDensity.current
                    val hc = com.microsoft.adaptivecards.rendering.theme.LocalHostConfig.current
                    val widthDp = with(density) { constraints.maxWidth.toDp().value }
                    val widthCategory = WidthCategory.fromDp(
                        widthDp,
                        veryNarrowBreakpoint = hc.hostWidth.veryNarrow,
                        narrowBreakpoint = hc.hostWidth.narrow,
                        standardBreakpoint = hc.hostWidth.standard
                    )

                    CompositionLocalProvider(
                        LocalWidthCategory provides widthCategory,
                        LocalCardViewModel provides viewModel,
                        LocalFeatureFlags provides configuration.featureFlags
                    ) {
                        Column(
                            modifier = Modifier
                                .fillMaxWidth()
                                .verticalScroll(rememberScrollState())
                        ) {
                            adaptiveCard.body?.forEachIndexed { index, element ->
                                RenderElement(
                                    element = element,
                                    isFirst = index == 0,
                                    viewModel = viewModel,
                                    actionHandler = actionHandler
                                )
                            }

                            adaptiveCard.actions?.let { actions ->
                                if (actions.isNotEmpty()) {
                                    ActionSetView(
                                        actions = actions,
                                        actionHandler = actionHandler,
                                        viewModel = viewModel,
                                        modifier = Modifier.fillMaxWidth()
                                    )
                                }
                            }
                        }
                    }
                }
            }
        }
    }
}

/**
 * Renders a single card element based on its type
 */
@Composable
fun RenderElement(
    element: CardElement,
    isFirst: Boolean = false,
    viewModel: CardViewModel,
    actionHandler: ActionHandler,
    modifier: Modifier = Modifier
) {
    // Check visibility
    if (!element.isVisible) return

    val elementId = element.id
    if (elementId != null && !viewModel.isElementVisible(elementId)) {
        return
    }

    // Check requires against host feature flags
    val featureFlags = LocalFeatureFlags.current
    if (!featureFlags.meetsRequirements(element.requires)) {
        // Requirements not met — render fallback or nothing
        val fallbackElement = resolveFallback(element.fallback)
        if (fallbackElement != null && fallbackElement !is UnknownElement) {
            RenderElement(fallbackElement, isFirst, viewModel, actionHandler, modifier)
        } else if (fallbackElement is UnknownElement && fallbackElement.unknownType == "drop") {
            // "drop" — render nothing
        } else if (fallbackElement != null) {
            RenderElement(fallbackElement, isFirst, viewModel, actionHandler, modifier)
        }
        return
    }

    // Check targetWidth constraint
    val widthCategory = LocalWidthCategory.current
    if (!shouldShowForTargetWidth(element.targetWidth, widthCategory)) {
        return
    }

    Column(modifier = modifier) {
        // Render separator
        if (element.separator && !isFirst) {
            SeparatorLine()
        }

        // Apply spacing and render element
        val elementModifier = Modifier.adaptiveSpacing(element.spacing, isFirst)

        when (element) {
            is TextBlock -> TextBlockView(element, elementModifier)
            is Image -> ImageView(element, elementModifier, actionHandler)
            is Container -> ContainerView(element, elementModifier, viewModel, actionHandler)
            is ColumnSet -> ColumnSetView(element, elementModifier, viewModel, actionHandler)
            is FactSet -> FactSetView(element, elementModifier)
            is ImageSet -> ImageSetView(element, elementModifier, actionHandler)
            is ActionSet -> ActionSetView(element.actions, actionHandler, viewModel, elementModifier)
            is RichTextBlock -> RichTextBlockView(element, elementModifier, actionHandler)
            is Media -> MediaView(element, elementModifier)
            is Table -> TableView(element, elementModifier, viewModel, actionHandler)
            is Icon -> IconView(element, elementModifier, actionHandler)
            is Badge -> BadgeView(element, elementModifier)
            // Input elements - check registry for host-provided renderers first
            is InputText, is InputNumber, is InputDate, is InputTime,
            is InputToggle, is InputChoiceSet, is InputDataGrid, is RatingInput -> {
                val inputRenderer = GlobalElementRendererRegistry.getRenderer(element.type)
                if (inputRenderer != null) {
                    inputRenderer(element, elementModifier)
                } else {
                    // Fallback placeholder — register input renderers via
                    // GlobalElementRendererRegistry to replace this
                    androidx.compose.material3.Text(
                        text = "[Input: ${element::class.simpleName}]",
                        modifier = elementModifier,
                        color = androidx.compose.material3.MaterialTheme.colorScheme.outline
                    )
                }
            }
            // Advanced elements
            is Carousel -> CarouselView(element, viewModel, actionHandler, elementModifier)
            is Accordion -> AccordionView(element, viewModel, actionHandler, elementModifier)
            is CodeBlock -> CodeBlockView(element, elementModifier)
            is RatingDisplay -> RatingDisplayView(element, elementModifier)
            is ProgressBar -> ProgressBarView(element, elementModifier)
            is ProgressRing -> ProgressRingView(element, elementModifier)
            is Spinner -> SpinnerView(element, elementModifier)
            is TabSet -> TabSetView(element, viewModel, actionHandler, elementModifier)
            is ListElement -> ListView(element, viewModel, actionHandler, elementModifier)
            is CompoundButton -> CompoundButtonView(element, actionHandler, elementModifier)
            // Chart elements
            is BarChart -> BarChartView(chart = element)
            is DonutChart -> DonutChartView(chart = element)
            is LineChart -> LineChartView(chart = element)
            is PieChart -> PieChartView(chart = element)
            else -> {
                // Check custom element registry for host-registered renderers
                val customRenderer = GlobalElementRendererRegistry.getRenderer(element.type)
                if (customRenderer != null) {
                    customRenderer(element, elementModifier)
                } else if (element is UnknownElement) {
                    // Unknown element type — try to render fallback
                    val fallbackElement = resolveFallback(element.fallback)
                    if (fallbackElement != null && !(fallbackElement is UnknownElement && fallbackElement.unknownType == "drop")) {
                        RenderElement(fallbackElement, isFirst, viewModel, actionHandler, elementModifier)
                    }
                    // else: no fallback or "drop" — render nothing
                }
            }
        }
    }
}

/**
 * Resolves a fallback JsonElement into a CardElement for rendering.
 * Handles both the "drop" string shorthand and full element objects.
 * Returns null if the fallback is null or cannot be parsed.
 */
private fun resolveFallback(fallback: kotlinx.serialization.json.JsonElement?): CardElement? {
    if (fallback == null) return null
    return try {
        // Handle "drop" string shorthand
        if (fallback is kotlinx.serialization.json.JsonPrimitive && fallback.isString) {
            val value = fallback.content
            if (value.equals("drop", ignoreCase = true)) {
                return UnknownElement(unknownType = "drop")
            }
            return null
        }
        // Parse full element object
        kotlinx.serialization.json.Json { ignoreUnknownKeys = true }
            .decodeFromJsonElement(com.microsoft.adaptivecards.core.parsing.CardElementSerializer, fallback)
    } catch (_: Exception) {
        null
    }
}

/**
 * Recursively collects all actions from a card (card-level + body ActionSets).
 */
fun collectAllActionsForCard(card: AdaptiveCard): List<CardAction> {
    val actions = mutableListOf<CardAction>()
    card.actions?.let { actions.addAll(it) }
    card.body?.let { body -> collectActionsFromElements(body, actions) }
    return actions
}

private fun collectAllActions(card: AdaptiveCard?): List<CardAction> {
    if (card == null) return emptyList()
    return collectAllActionsForCard(card)
}

private fun collectActionsFromElements(elements: List<CardElement>, actions: MutableList<CardAction>) {
    for (element in elements) {
        if (element is ActionSet) {
            actions.addAll(element.actions)
        }
        if (element is Container) {
            element.items?.let { collectActionsFromElements(it, actions) }
        }
        if (element is ColumnSet) {
            element.columns?.forEach { column ->
                column.items?.let { collectActionsFromElements(it, actions) }
            }
        }
    }
}
